File: backend/src/models/productModel.js
// backend/src/models/productModel.js
const mongoose = require('mongoose');
const {model} = require("mongoose");
const InventoryModel = require("./inventoryModel");

const productSchema = new mongoose.Schema({

    name: {
        type: String,
        required: true,
    },

    price:{
        type: Number,
        required: true,
    },

    inventory: {
        type: Number,
        default: 1,
    },

    targetInventory: {
        type: Number,
        default: 50,
    },

    categoryID: {
        type:mongoose.Schema.Types.ObjectId,
        ref: 'Category',
    },

    images:{
        type: Array,
    },

    isOnSale:{
        type: Boolean,
        default: false,
    }

});


//region✅ add inventory for new product method
productSchema.pre("save", async function (next) {

    try {
        const productID = this._id;
        const inventory = this.inventory;
        const  targetInventory = this.targetInventory;

        const newInventoryForNewProduct = await new InventoryModel({
            productID,
            inventory,
            targetInventory
        });

        await newInventoryForNewProduct.save();

        // next();

    } catch (e) {
        console.log(e);
    }

});
//endregion

const ProductModel = model("Products", productSchema);
module.exports.ProductModel = ProductModel;

File: flutter_app/lib/app/data/models/product_model.dart
import 'package:flutter_app/app/data/models/category_model.dart';

class ProductModel {
  String name;
  double price;
  double inventory;
  double targetInventory;
  CategoryModel categoryModel;
  List<String> images;
  bool? isOnSale = false;

  ProductModel({
    required this.name,
    required this.price,
    required this.inventory,
    required this.targetInventory,
    required this.categoryModel,
    required this.images,
    this.isOnSale = false,
  });

  factory ProductModel.fromJson(Map<String, dynamic> json) {
    return ProductModel(
      name: json['name'] ?? "",
      price: json['price'] ?? 0,
      inventory: json['inventory'] ?? 0,
      targetInventory: json['targetInventory'] ?? 0,
      categoryModel: CategoryModel.fromJson(json['category'] ?? {}),
      images: json['images'] ?? [],
      isOnSale: json['isOnSale'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'price': price,
      'inventory': inventory,
      'targetInventory': targetInventory,
      'category': categoryModel.toJson(),
      'images': images,
      'isOnSale': isOnSale,
    };
  }
}
